### PR TITLE
add -P (preserve permission) option to reveal and hide. For #172

### PR DESCRIPTION
**Pull request by joshrabinowitz**
For #172. This adds a -P option to 'git secret reveal' like cp's -p (preserve) option.

(-p is already used for 'password' by git-secret)

Suggestions on code changes or specific tests to add welcome.

@sobolevn @gthank  @simbo1905

### CODE REVIEW

**Comment by gthank** [comment]
This looks awesome. Thanks a ton! I still wish it was the default, but just having the option is enough to save me a ton of work, so really: thanks a lot for this work.

**Comment by joshrabinowitz** [comment]
@gthank you're welcome! Yeah, we can't change default behavior until @sobolevn desires it and we reach 0.3.0.

I'm not super-happy using -c instead of -p, suggestions welcome. 

Also, if `git-secret reveal -c` cannot set the permissions of the decrypted file, and should it be a warning, or an error?

**Comment by joshrabinowitz** [comment]
Also, should there be a similar setting on 'git-secret hide' to set the perms of the hidden file to match the perms of the uncompressed one?

**Comment by gthank** [comment]
I'm inclined to say it should be an error. If people are relying on it to set the permissions, and then it doesn't, things are likely to break later, and be harder to spot.

As for `-c` vs `-p`, I'm not sure. If the convention is to always have a "short" version, then `-p` is already taken, and "preserve permissions" doesn't really offer an obvious alternative. `-c` for `--chmod` is as good a mnemonic as any I can think of at the moment. If you want to only provide the "long" form, then I'd be fine with using `--preserve`.

**Comment by gthank** [comment]
And yes, a parallel flag on either `add` or `hide` to preserve the permissions would be ideal.

**Comment by joshrabinowitz** [comment]
Re: a `--preserve` option: at this point all of git-secret uses short (single-letter) options.

**Comment by joshrabinowitz** [comment]
I should get around to finishing this next week.

File: src/_utils/_git_secret_tools_linux.sh [hunk 1]
```diff
@@ -16,3 +16,11 @@ function __temp_file_linux {
 function __sha256_linux {
   sha256sum "$1"
 }
+
```
Comment (joshrabinowitz) [inline src/_utils/_git_secret_tools_linux.sh]: @sobolevn @simbo1905 

I'm welcoming input on the way to implement the -c option and get the 'permission string' to use for chmod? 

* I figured octal was the best representation to use. 
Agreed?

* I couldn't find just one incantation that would return the octal permissions on OSX and linux, so I'm using `stat --format '%a'` on linux and `stat -f '%p'` on osx. 
Let me know if you see a better way.

File: src/_utils/_git_secret_tools_linux.sh [hunk 2]
```diff
@@ -16,3 +16,11 @@ function __temp_file_linux {
...
+function __get_octal_perms_linux {
+  local filename
+  filename=$1
+  local perms
+  perms=$(stat --format '%a' "$filename")
```
Comment (joshrabinowitz) [inline src/_utils/_git_secret_tools_linux.sh]: Here we use `stat --format '%a'`

File: src/_utils/_git_secret_tools_osx.sh [hunk 3]
```diff
@@ -18,3 +18,10 @@ function __temp_file_osx {
 function __sha256_osx {
   /usr/bin/shasum -a256 "$1"
 }
+function __get_octal_perms_osx {
+  local filename
+  filename=$1
+  local perms
+  perms=$(stat -f '%p' "$filename")
```
Comment (joshrabinowitz) [inline src/_utils/_git_secret_tools_osx.sh]: Here we use `stat -f '%p'`

**Comment by joshrabinowitz** [comment]
I think this is ready for review and/or merging, @gthank @simbo1905 @sobolevn

**Pull request merged**